plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java

```diff
@@ -0,0 +1,121 @@
+/*******************************************************************************
+ * Copyright (c) 2015 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v1.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package org.jboss.tools.openshift.internal.core;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.jboss.tools.openshift.common.core.connection.ConnectionsRegistry;
+import org.jboss.tools.openshift.common.core.connection.ConnectionsRegistrySingleton;
+import org.jboss.tools.openshift.core.connection.Connection;
+import org.jboss.tools.openshift.core.connection.ConnectionProperties;
+import org.jboss.tools.openshift.core.connection.ConnectionsRegistryUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.openshift.restclient.IOpenShiftWatchListener;
+import com.openshift.restclient.IWatcher;
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.capability.CapabilityVisitor;
+import com.openshift.restclient.capability.resources.IClientCapability;
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.IResource;
+
+public class WatchManager {
+
+	private static final Collection<String> WATCH_KINDS = Arrays.asList(ResourceKind.BUILD, ResourceKind.POD);
+	private static final String DELIMITER = "|";
+	private static WatchManager instance;
+	private static final Logger LOGGER = LoggerFactory.getLogger(WatchManager.class);
+	
+	private Map<String, IWatcher> watches = Collections.synchronizedMap(new HashMap<>());
+
+	public static WatchManager getInstance() {
```

Comment:
either use `synchronized` or better, the Initialization On Demand Holder pattern : http://blog.crazybob.org/2007/01/lazy-loading-singletons.html